apm-sniffer/apm-sdk-plugin/hystrix-1.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/hystrix/v1/SWHystrixLifecycleForwardingRequestVariable.java

```diff
@@ -0,0 +1,84 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.hystrix.v1;
+
+import com.netflix.hystrix.strategy.concurrency.HystrixRequestContext;
+import com.netflix.hystrix.strategy.concurrency.HystrixRequestVariable;
+import com.netflix.hystrix.strategy.concurrency.HystrixRequestVariableDefault;
+import com.netflix.hystrix.strategy.concurrency.HystrixRequestVariableLifecycle;
+import com.netflix.hystrix.strategy.concurrency.HystrixConcurrencyStrategy;
+
+/**
+ * Implementation of {@link HystrixRequestVariable} which forwards to the wrapped
+ * {@link HystrixRequestVariableLifecycle}.
+ * <p>
+ * This implementation also returns null when {@link #get()} is called while the {@link HystrixRequestContext} has not
+ * been initialized rather than throwing an exception, allowing for use in a {@link HystrixConcurrencyStrategy} which
+ * does not depend on an a HystrixRequestContext
+ */
+public class SWHystrixLifecycleForwardingRequestVariable<T> extends HystrixRequestVariableDefault<T> {
+    private final HystrixRequestVariableLifecycle<T> lifecycle;
+
+    /**
+     * Creates a HystrixRequestVariable which will return data as provided by the {@link HystrixRequestVariableLifecycle}
+     *
+     * @param lifecycle lifecycle used to provide values. Must have the same type parameter as the constructed instance.
+     */
+    public SWHystrixLifecycleForwardingRequestVariable(HystrixRequestVariableLifecycle<T> lifecycle) {
+        this.lifecycle = lifecycle;
+    }
+
+    /**
+     * Delegates to the wrapped {@link HystrixRequestVariableLifecycle}
+     *
+     * @return T with initial value or null if none.
+     */
+    @Override
+    public T initialValue() {
+        return lifecycle.initialValue();
+    }
+
+    /**
+     * Delegates to the wrapped {@link HystrixRequestVariableLifecycle}
+     *
+     * @param value of request variable to allow cleanup activity.
+     *              <p>
+     *              If nothing needs to be cleaned up then nothing needs to be done in this method.
+     */
+    @Override
+    public void shutdown(T value) {
+        lifecycle.shutdown(value);
+    }
+
+    /**
+     * Return null if the {@link HystrixRequestContext} has not been initialized for the current thread.
+     * <p>
+     * If {@link HystrixRequestContext} has been initialized then call method in superclass:
+     * {@link HystrixRequestVariableDefault#get()}
+     */
+    @Override
+    public T get() {
+        if (!HystrixRequestContext.isCurrentThreadInitialized()) {
+            return null;
+        }
+        return super.get();
```

Comment:
So, this could be NPE, right? In some old versions.

Reply:
this is super.get() method
```
public T get() {
        if (HystrixRequestContext.getContextForCurrentThread() == null) {
            throw new IllegalStateException(HystrixRequestContext.class.getSimpleName() + ".initializeContext() must be called at the beginning of each request before RequestVariable functionality can be used.");
        } else {
            ConcurrentHashMap<HystrixRequestVariableDefault<?>, HystrixRequestVariableDefault.LazyInitializer<?>> variableMap = HystrixRequestContext.getContextForCurrentThread().state;
            HystrixRequestVariableDefault.LazyInitializer<?> v = (HystrixRequestVariableDefault.LazyInitializer)variableMap.get(this);
            if (v != null) {
                return v.get();
            } else {
                HystrixRequestVariableDefault.LazyInitializer<T> l = new HystrixRequestVariableDefault.LazyInitializer(this);
                HystrixRequestVariableDefault.LazyInitializer<?> existing = (HystrixRequestVariableDefault.LazyInitializer)variableMap.putIfAbsent(this, l);
                return existing == null ? l.get() : existing.get();
            }
        }
    }
```
super.get() will trigger IllegalStateException, in some old versions

Reply:
Then next question is, why this method is being called? Is it called by SkyWalking plugin codes?

Reply:
it ic called by hystrix AbstractCommand constructor.

Reply:
My question is, the old version caller, AbstractCommand, should not call this, right? Otherwise, it triggers Illegal issue itself. Could you be more clear what context is missing

Reply:
concurrencyStrategy is HystrixConcurrencyStrategyDefault instance in no skywalking agent, but skywalking plugin wrapped it to HystrixConcurrencyStrategy instance. not HystrixConcurrencyStrategyDefault. thus skip "concurrencyStrategy instanceof HystrixConcurrencyStrategyDefault" condition. i don't know where context be initialize, but when "concurrencyStrategy instanceof HystrixConcurrencyStrategyDefault" condition, it checkd  this "HystrixRequestContext.isCurrentThreadInitialized()".
![image](https://user-images.githubusercontent.com/16839929/83763364-32325800-a6ab-11ea-9cc6-3073159f6838.png)

Reply:
Then, why don't we change the SkyWalking wrapper logic? It seems more reasonable, right?

Reply:
With offline communication, right now, this seems the easiest solution w/o harm. I will keep this for now. Thanks.